### PR TITLE
Add support for Vector2i, Rect2i and Vector3i to Variant

### PR DESCRIPTION
**Pull request by reduz**
**EDIT** WARNING: Requires C++17 'guaranteed copy elision' in order to solve ambiguities in variant. See comments below.

~~WARNING: Requires C++17, I just can't make this work with
C++14. I fall into the trouble we always had before of
ambiguous operators, which is too complex to fix by using
explicit keywords, even with C++14.~~

~~It seems C++17 has saner rules, prioritizing exact match to
disambiguate. Trying to solve this in C++14 results in large
amount of hacks.~~

### CODE REVIEW

**Comment by reduz** [comment]
If you want to  investigate this problem further, here is a minimal reproduction case. C++14 fails with ambiguous overload, while C++17 works fine using best match.
```


struct A {

    int x = 0;    
};


struct B {

    B() {}
    B(const B& p_b) {} 
    B(const A& p_a) {}
    operator A() const { A a; return a; }    
};


struct C {

    C() {}
    C(const A& p_a) {}
    operator A() { A a; return a; }    
    operator B() const { B b; return b; }    
};


int main() {

    C c;
    B a(c);

    return 0;
}
```

Reproduced in both gcc and clang for the 14 and 17 versions. I wish I could find the relevant section in the [C++17 specification](https://isocpp.org/files/papers/p0636r0.html) that makes it work there, but I just fail to locate it.

File: SConstruct [hunk 1]
```diff
@@ -316,11 +316,11 @@ if selected_platform in platform_list:
         # both GCC and Clang. This mirrors GCC and Clang's current default
```
Comment (YeldhamDev) [inline SConstruct]: This comment should probably be updated.

**Comment by reduz** [comment]
Ok, I've been doing my research. It appears that what makes this work in C++17 is something called "guaranteed copy elision":

`
if the initializer is a prvalue expression whose type is the same class as T (ignoring cv-qualification), the initializer expression itself, rather than a temporary materialized from it, is used to initialize the destination object.
(Before C++17, the compiler may elide the construction from the prvalue temporary in this case, but the appropriate constructor must still be accessible: see copy elision)
`

Here is a [link to stackoverflow[(https://stackoverflow.com/questions/38043319/how-does-guaranteed-copy-elision-work).

It would appear that the way to make this work in C++14 is via use of [move semantics](https://en.cppreference.com/w/cpp/language/move_constructor), while in C++17 its no longer necessary to achieve the same effect in this case.

**Comment by akien-mga** [comment]
@reduz: Build fails on macOS' Clang:
```
editor/plugins/path_editor_plugin.cpp:197:111: error: ambiguous conversion for static_cast from 'const Variant' to 'Vector3'
                        ur->add_undo_method(c.ptr(), "set_point_out", idx, PathEditorPlugin::singleton->mirror_length_enabled() ? -static_cast<Vector3>(p_restore) : (-static_cast<Vector3>(p_restore).normalized() * orig_out_length));
                                                                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
./core/math/vector3.h:40:8: note: candidate is the implicit move constructor
struct Vector3 {
       ^
./core/math/vector3.h:40:8: note: candidate is the implicit copy constructor
./core/math/vector3.h:155:17: note: candidate constructor
        _FORCE_INLINE_ Vector3(const Vector3i &p_ivec) {
                       ^
```
(4 errors on similar casts in that file)

**Comment by reduz** [comment]
Thats the error when you are not using C++17 I think, are you sure the flag is working?

**Comment by aaronfranke** [comment]
Related to https://github.com/godotengine/godot/pull/22205, but does not fully supercede since it's still not in C# (and maybe not VisualScript?).

@akien-mga I faced that same ambiguity and managed to work around the problem without C++17. If you're interested, see the first two commits in the history ([commit 1](https://github.com/godotengine/godot/pull/22205/commits/ca337d2f02433c540cc63ea077eafbeacd5405fc) and [commit 2](https://github.com/godotengine/godot/pull/22205/commits/ed27a4d1470ad75e785ace9e040ecb29d6b96db0)).

**Comment by akien-mga** [comment]
Might be that the Xcode version used (9.4.1) is too old, according to https://stackoverflow.com/a/57913542 I might need Xcode 10. I'll try to update Travis.

**Comment by akien-mga** [comment]
Confirmed that updating to Xcode 11.3 fixed it.
I'm pushing it as a separate commit in #36464, I'll rebase this once it's merged.

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Headsup: I found that while this builds fine with GCC 7 and GCC 9, it fails with GCC 8 which seems to have broken guaranteed copy elision. Reproduced both on Mageia 7 (GCC 8.3.1 snapshot from late January) and Ubuntu 18.04 with GCC 8.3.0.

I'll report it upstream.

**Pull request merged**